2(a)  (2 points) 
Using the CDS coordinates from 1(b), extract the coding sequence from the mRNA. Report its length in 
nucleotides and confirm that it is divisible by three. If it is not, stop and work out why before continuing — 
do not proceed with a sequence that fails this check. 
WHY: A coding sequence that is not a whole number of codons means your extraction is wrong, not that the 
gene is unusual. 

In [9]:
with open("MTCYB_CDS.fasta", "r") as fh: # open() opens the fasta file in the read mode and fh is a name assigned to this file
    New_sequence = fh.read().replace("\n", "") # Reading the entire file as a continuous string by removing the white lines
    New_sequence = New_sequence + "AA" 

with open("MTCYB_CDS_corrected.fasta", "w") as fh: # Creating a new file and adding the new sequence containing the double A residues to it
    fh.write(New_sequence)

print(len(New_sequence))
print(len(New_sequence) / 3)

1143
381.0


381 because 1 codon will code for the stop codon, thus logically the actual amino acid length should be 380

2(b)  (4 points) 
Build the genetic code as a lookup structure of your own making, then use it to translate the entire coding 
sequence to protein. Report the first 30 residues of your translation and total length. 
You must construct the lookup yourself. You may read the codon assignments from a published table, but 
the data structure and the translation loop must be your own work. 
Hint: A dictionary keyed on the three-letter codon is the natural structure — Lab Session 5, Exercise 4.4. Sixty
four entries written out in full is perfectly acceptable and easier to check than anything clever. 

In [10]:
Table = {
    "TTT":"F", "TCT":"S", "TAT":"Y", "TGT":"C",
    "TTC":"F", "TCC":"S", "TAC":"Y", "TGC":"C",
    "TTA":"L", "TCA":"S", "TAA":"", "TGA":"W",
    "TTG":"L", "TCG":"S", "TAG":"", "TGG":"W",
    "CTT":"L", "CCT":"P", "CAT":"H", "CGT":"R",
    "CTC":"L", "CCC":"P", "CAC":"H", "CGC":"R",
    "CTA":"L", "CCA":"P", "CAA":"Q", "CGA":"R",
    "CTG":"L", "CCG":"P", "CAG":"Q", "CGC":"R",
    "ATT":"I", "ACT":"T", "AAT":"N", "AGT":"S",
    "ATC":"I", "ACC":"T", "AAC":"N", "AGC":"S",
    "ATA":"M", "ACA":"T", "AAA":"K", "AGA":"",
    "ATG":"M", "ACG":"T", "AAG":"K", "AGG":"",
    "GTT":"V", "GCT":"A", "GAT":"D", "GGT":"G",
    "GTC":"V", "GCC":"A", "GAC":"D", "GGC":"G",
    "GTA":"V", "GCA":"A", "GAA":"E", "GGA":"G",
    "GTG":"V", "GCG":"A", "GAG":"E", "GGG":"G"
}
with open("MTCYB_CDS_corrected.fasta", "r") as fh: # open() openes the fasta file in the read mode and fh is a name assigned to this file
    seq = fh.read().replace("\n", "") # Reading the entire file as a continuous string by removing the white lines
Translated_protein_seq = "" # Creating an empty string to add amino acid residues to it
for i in range(0, len(seq), 3): # A for loop is initiated with index 0 to maximum index being the length of the sequence and number of steps being 3 as each codon consists of 3 bases
    codon = seq[i:i+3] # 3 nucleotides from seq is taken and assigned to a variable named codon which acts as a key in the dictionary
    AA_residue = Table[codon] #The key is used to extract the values from table and assign it to a variable named AA_residue
    Translated_protein_seq = Translated_protein_seq + AA_residue # Amino acids are added to the protein sequence 
print (Translated_protein_seq)
print("First 30 residues:", Translated_protein_seq[:30]) # Printing the first 30 amino acid residues
print("Total length of the protein sequence:", len(Translated_protein_seq)) # Calculating the total number of amino acids present in the protein sequence

MTPMRKTNPLMKLINHSFIDLPTPSNISAWWNFGSLLGACLILQITTGLFLAMHYSPDASTAFSSIAHITRDVNYGWIIRYLHANGASMFFICLFLHIGRGLYYGSFLYSETWNIGIILLLATMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTDLVQWIWGGYSVDSPTLTRFFTFHFILPFIIAALATLHLLFLHETGSNNPLGITSHSDKITFHPYYTIKDALGLLLFLLSLMTLTLFSPDLLGDPDNYTLANPLNTPPHIKPEWYFLFAYTILRSVPNKLGGVLALLLSILILAMIPILHMSKQQSMMFRPLSQSLYWLLAADLLILTWIGGQPVSYPFTIIGQVASVLYFTTILILMPTISLIENKMLKWA
First 30 residues: MTPMRKTNPLMKLINHSFIDLPTPSNISAW
Total length of the protein sequence: 380


2(c)  (1 points) 
Compare your translation to the deposited protein from 1(c). Report whether the two are identical. If they 
are not, report the position of the first difference and what each sequence has at that position. State the 
tool you used to compare them. 
Hint: Strip the FASTA header and remove all line breaks from both sequences before comparing. A difference 
caused by line wrapping is not a difference in the protein.

Answer: Header and line breaks were previously removed from the fasta files

In [11]:
with open("deposited_protein.fasta","r") as fh:
    deposited_protein = fh.read() 
seq_1 = Translated_protein_seq # Contains the translated protein sequence
seq_2 = deposited_protein # Contains the deposited protein sequence
print(len(Translated_protein_seq))
print(len(deposited_protein))
differences = 0
for a, b in zip(seq_1, seq_2): # zip aligns seq 1 and seq 2
    if a != b:
        differences = differences + 1 # Every time a different amino acid is encountered, the counter is increased by one
print(differences)

380
380
0


The 2 sequences are identical

2(d)  (1 point) 
Cross-check your own translation against a library function — Biopython's translate, or any equivalent. 
Confirm that the two agree, or explain exactly where and why they do not. 
NOTE: A new tool is only trustworthy once you have checked it against one you already trust. 

In [13]:
from Bio.Seq import Seq
with open("MTCYB_CDS_corrected.fasta", "r") as fh:
    seq = fh.read().replace("\n", "")
seq = Seq(seq) # Conversion of python string to Biopython Seq object
Translated_protein_sequence = seq.translate(table = 2) # Translating the sequence to a protein sequence.transl_table=2 mentioned in cds of MT-CYB gene in genbank record
print("Length of Biopythontranslated sequence:", len(Translated_protein_sequence))
print("Biopython translated sequence:", Translated_protein_sequence)
print("Length of my translated sequence:", len(Translated_protein_seq))
print("My translated sequence:", Translated_protein_seq)
seq_1 = Seq(Translated_protein_sequence)
seq_2 = Seq(Translated_protein_seq)
differences = 0
for a, b in zip(seq_1, seq_2):
    if a != b:
        differences = differences + 1 # Every time a different amino acid is encountered, the counter is increased by one
print("Number of differences in amino acid residues:", differences) 

Length of Biopythontranslated sequence: 381
Biopython translated sequence: MTPMRKTNPLMKLINHSFIDLPTPSNISAWWNFGSLLGACLILQITTGLFLAMHYSPDASTAFSSIAHITRDVNYGWIIRYLHANGASMFFICLFLHIGRGLYYGSFLYSETWNIGIILLLATMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTDLVQWIWGGYSVDSPTLTRFFTFHFILPFIIAALATLHLLFLHETGSNNPLGITSHSDKITFHPYYTIKDALGLLLFLLSLMTLTLFSPDLLGDPDNYTLANPLNTPPHIKPEWYFLFAYTILRSVPNKLGGVLALLLSILILAMIPILHMSKQQSMMFRPLSQSLYWLLAADLLILTWIGGQPVSYPFTIIGQVASVLYFTTILILMPTISLIENKMLKWA*
Length of my translated sequence: 380
My translated sequence: MTPMRKTNPLMKLINHSFIDLPTPSNISAWWNFGSLLGACLILQITTGLFLAMHYSPDASTAFSSIAHITRDVNYGWIIRYLHANGASMFFICLFLHIGRGLYYGSFLYSETWNIGIILLLATMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTDLVQWIWGGYSVDSPTLTRFFTFHFILPFIIAALATLHLLFLHETGSNNPLGITSHSDKITFHPYYTIKDALGLLLFLLSLMTLTLFSPDLLGDPDNYTLANPLNTPPHIKPEWYFLFAYTILRSVPNKLGGVLALLLSILILAMIPILHMSKQQSMMFRPLSQSLYWLLAADLLILTWIGGQPVSYPFTIIGQVASVLYFTTILILMPTISLIENKMLKWA
Number of differences in amino acid residues: 0
